### PR TITLE
Lauren pipes MediaRanker

### PR DESCRIPTION
**Pull request by laurenelee**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | In my Users_controller, I wrote a login method that allowed the user to sign in as an existing user. It got a little complicated, so I pulled out a nested if look from the login method to make a create method. That method created a new user. 
Describe how you approached testing that model method. What edge cases did you come up with? | My create method required a session logged in user and so I had to create that login session running the post login_path to sign in before testing the redirect location.
Describe an edge case test you wrote for a controller | In the Users_controller, if the user does not exist, the error status :not_found is generated. 
What are `session` and `flash`? What is the difference between them? | Session is used to store info about a user when they login. Rails auto deletes the info when user closes the browser. Similar to params but different because it persists from one request to another. Flash is a special type of hash used to send one-time messages from controllers to our views. Used to denote success and failure moments in the app. 
Describe a controller filter you wrote. | I created a private method called find_work and ran the before action for :show, :edit, :update, :destroy, which helped the controlled DRY up quite a bit. 
What was one thing that you gained more clarity on through this assignment? | Testing! Although I certainly would like to learn more about testing sessions. 
What is the Heroku URL of your deployed application | https://lauren-mediaranker.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? | Perhaps swapping this one with rideshare? It might be interesting as a pair project whereas I felt that rideshare was doable as a solo assignment.

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  root 'main#index'
+  ####### main ######
+
+  ####### works ######
+  resources :works
+  ####### works ######
+
+  ####### users and login ######
+  get 'users', to: 'users#index', as: 'users'
+  get 'users/:id', to: 'users#show', as: 'user'
```
Comment (droberts-sea) [inline config/routes.rb]: This is fine, but you could also have said:

```ruby
resources :users, only: [:index, :show]
```

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      flash[:status] = :failure
+      flash[:message] = "You are already logged in as #{User.find_by(id: session[:user_id].to_i).name}"
+      redirect_to root_path
+    end
+  end
+
+  def create
+    unless params {name.blank?}
+      if User.create(name: params[:name])
+        user = User.find_by(name: params[:name])
```
Comment (droberts-sea) [inline app/controllers/users_controller.rb]: The name of this method confused me at first - `create` is usually a controller action, but here you have it as a helper method. I do like that you've extracted this functionality, but it might be wise to rename it to something like `build_new_user` and mark it as `private`.

File: app/controllers/works_controller.rb [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  def index
+    @works = Work.all
+
+    @movies = Work.sort_by_category("movie")
+    @books = Work.sort_by_category("book")
+    @albums = Work.sort_by_category("album")
+    #
+    unless @works
+      # how to trigger this and make this actually occur?
+      head :not_found
```
Comment (droberts-sea) [inline app/controllers/works_controller.rb]: If there are no works, `Work.all` will return an empty array, which is truthy. Only `nil` or `false` will ever make this `unless` block happen.

However, I'm not sure `not_found` is what you want if there are no works. Wouldn't it make sense to just show 3 empty tables?

File: app/controllers/works_controller.rb [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  private
+
+  def work_params
+    return params.require(:work).permit(:category, :title, :creator, :publication_year, :description)
+  end
+
+  def find_work
+    @work = Work.find_by(id: params[:id])
+    unless @work
+      head :not_found
```
Comment (droberts-sea) [inline app/controllers/works_controller.rb:82]: Good use of a controller filter here to DRY up this controller.

File: app/models/work.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
+class Work < ApplicationRecord
+  has_many :votes
+
+  validates :title, presence: true, uniqueness: { message: "The title was not unique! Please choose a different one" }
+  validates :description, length: { maximum: 500, message: "The description is too long" }
+
+  def self.sort_by_category(category)
+      return Work.where(category: category).sort_by{|work| -work.votes.count}
+  end
```
Comment (droberts-sea) [inline app/models/work.rb:9]: Good job moving logic into the model.

File: test/models/work_test.rb [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+      let :work { Work.first }
+
+      let :user1 { User.last }
+      let :work1 { Work.last }
+
+      let :user2 { User.first + 1  }
+
+      let :vote { Vote.new(work: work, user: user) }
+      let :vote1 { Vote.new(work: work1, user: user1) }
+      let :vote2 { Vote.new(work: work, user: user2) }
```
Comment (droberts-sea) [inline test/models/work_test.rb]: All of these `let` statements will not be executed unless the variables they create are referenced. This means that in this test, you won't have all those votes, which probably makes the test invalid.

Instead, you should just do this work at the top of the test block, as part of the Arrange step.

**Comment by droberts-sea** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | yes
Agile practices (trello board) | yes
Comprehension questions	| yes
**General** |
Rails fundamentals (RESTful routing, use of named paths) | yes
Semantic HTML | yes
Errors are reported to the user | yes
Business logic lives in the models | yes
Models are thoroughly tested, including relations, validations and any custom logic | yes
Controllers are thoroughly tested | yes - great work!
**Wave 1 - Media** |
Splash page shows the three media categories | yes
Basic CRUD operations on media are present and functional | yes
**Wave 2 - Users and Votes** |
Users can log in and log out | could not log in as a new user, but as an existing user worked fine
The ID of the current user is stored in the session | yes
Individual user pages and the user list are present | yes
A user cannot vote for the same media more than once | yes
All media lists are ordered by vote count | yes
Splash page contains a media spotlight | yes
Media pages contain lists of voting users | yes
**Wave 3 - Styling** |
Foundation is used appropriately | yes
Look and feel is similar to the original | yes
**Overall** | Excellent job overall! I've got a few inline comments below for you to review, but in general I am quite happy with this submission. Keep up the hard work!
